src/hash.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2018 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/hash.h"
+#include "mococrw/util.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw
+{
+using namespace openssl;
+
+static const std::string SHA256 = "sha256";
+static const std::string SHA512 = "sha512";
+
+std::vector<uint8_t> sha256(const uint8_t* message, int length) {
+   return Hash::sha256().update(message, length).digest();
+}
+
+std::vector<uint8_t> sha256(const std::string &message) {
+    return Hash::sha256().update(message).digest();
+}
+
+std::vector<uint8_t> sha256(const std::vector<uint8_t> &message) {
+    return Hash::sha256().update(message).digest();
+}
+
+std::vector<uint8_t> sha512(const uint8_t* message, int length) {
+    return Hash::sha512().update(message, length).digest();
+}
+
+std::vector<uint8_t> sha512(const std::string &message) {
+    return Hash::sha512().update(message).digest();
+}
+
+std::vector<uint8_t> sha512(const std::vector<uint8_t> &message) {
+    return Hash::sha512().update(message).digest();
+}
+
+std::map<DigestTypes, int> Hash::lengthBytes = {
+    { DigestTypes::SHA256, 256 / 8 },
+    { DigestTypes::SHA512, 512 / 8 }
+};
+
+Hash::Hash(const DigestTypes digestType) : _digestType(digestType) {
+    const EVP_MD* digestFn = _getMDPtrFromDigestType(digestType);
+    _digestCtx = _EVP_MD_CTX_create();
+    lib::OpenSSLLib::SSL_EVP_MD_CTX_init(_digestCtx.get());
+    lib::OpenSSLLib::SSL_EVP_DigestInit_ex(_digestCtx.get(), digestFn, NULL);
+}
+
+Hash Hash::sha256 () {
+    return Hash{DigestTypes::SHA256};
+}
+
+Hash Hash::sha512 () {
+    return Hash{DigestTypes::SHA512};
+}
+
+std::vector<uint8_t> Hash::digest () {
+    if (_finalDigestValue.empty()) {
+        unsigned int length;
+        _finalDigestValue.resize(getDigestLengthBytes());
+        lib::OpenSSLLib::SSL_EVP_DigestFinal_ex(_digestCtx.get(), _finalDigestValue.data(), &length);
```

Comment:
Please do not call openssl functions from the code directly. Please define a wrapper that does the necessary error-handling in openssl_wrap.c. See other functions or ask on how this is usually done with this callChecked templates.

This way you just ignore the return value and do no error handling in case of openssl errors. One of the major goals of this libraries to get rid of this kind of behavior.

Reply:
I created basic wrappers. Do you think any custom error handling inside the wrappers would make sense here, e.g. when update is called after finalizing